src/packages/api/api_manager.py

```diff
@@ -0,0 +1,468 @@
+
+        try:
+            data = await self._call(f"rats/{uuid}")
+            return APIManager._deserialize_rat(data['data'])
+        except aiohttp.ClientError:
+            return None
+
+    async def find_rat(self,
+                       name: str,
+                       platform: typing.Optional[Platforms] = None) -> typing.Optional[Rat]:
+        """
+        Use the API to find a rat by name, and optionally platform.
+
+        Args:
+            name (str): The Rat's name.
+            platform (Platforms): The Rat's platform.
+
+        Returns:
+            A RatData object describing the Rat found, or None if none was found.
+        """
+        if not name:
+            raise ValueError("name must be specified!")
+
+        params = {'filter[name:ilike]': name}
+        if platform:
+            params['filter[platform:eq]'] = platform.name.casefold()
+
+        try:
+            data = await self._call('rats', params)
+            if data:
+                return APIManager._deserialize_rat(data['data'][0])
+        except aiohttp.ClientError:
+            return None
+
+    @staticmethod
+    def _build_uri_params(params: typing.Optional[typing.Dict[str, str]]) -> str:
+        """
+        Build a URI query string out of a Dict of key/value pairs.
+
+        Args:
+            params (Dict): The key/value pairs to build into a query.
+
+        Returns:
+            A properly HTML-escaped URI query string with leading question mark.
+        """
+        param_string = ""
+        if params:
+            param_string = '&'.join(
+                [f"{key}={escape(str(value))}" for key, value in params.items()]
+            )
+        return f"?{param_string}"
+
+    async def _call(self,
+                    endpoint: str,
+                    params: typing.Optional[typing.Dict[str, str]] = None,
+                    method: typing.Optional[str] = "GET",
+                    body: typing.Optional[typing.Dict] = None
+                    ) -> typing.Optional[typing.Union[dict, list]]:
+        """
+        Perform an API call on the Fuel Rats API.
+
+        Args:
+            endpoint (str): The API endpoint to request.
+            params (Dict): A dictionary of key-value pairs that will make up the query string.
+            method (str): The HTTP method to use for the API call.
+            body (Dict): The body of the request. (For HTTP POST/PATCH/PUT)
+
+        Returns:
+            A dict or list object representing the parsed JSONAPI data returned.
+            Returns None if in offline mode or the request fails.
+        """
+
+        if not self._online_mode:
+            LOG.debug(f"API call to {endpoint} was requested, but we're in offline mode.")
+            return None
+
+        base_url = self._base_url
+        param_string = self._build_uri_params(params)
+        url = f"{base_url}{endpoint}{param_string}"
+        for retry in range(self.MAX_RETRIES):
```

Comment:
a routine like this may be useful in other modules, perhaps we can extract this as an async generator?

Reply:
(for clarity, im refering to the retry loop)